### DISCUSSION TITLE
Evaluate Firebase Emulator Suite for Tests

### DISCUSSION THREAD

**Discussion by jeromegamez** [kreait/firebase-php#456] [Ideas]
https://firebase.google.com/docs/emulator-suite

Source: http://firebase.googleblog.com/2020/07/product-news-and-other-highlights-from.html

**Comment by nicholasio** [comment]
@jeromegamez any suggestions for where to start? I'm interested in working out a PR to get this supported added.

**Comment by jeromegamez** [comment]
Unfortunately the current setup of the SDK doesn't make it easy to integrate the emulator without rewriting large portions of the SDK - hardcoded URLs all over the place (in this case, when I say "the SDK is mature" I mean "it's old" 😅). The only way I see without a rewrite is to somehow use a Guzzle Middleware to intercept requests. I don't know if it's actually possible though 😅

**Comment by nicholasio** [comment]
for firestore it looks like all it's needed is setting the `FIRESTORE_EMULATOR_HOST` env variable.

I have a project using the non-gRPC client (firestore-php) and it worked fine with a small change: https://github.com/nicholasio/firestore-php/commit/77420c68ed8561056204a2214aae58468cca962b

I looked at the official PHP library for firestore and it also seems to support that env variable already.  

The missing piece is making this lib work with firebase auth emulator. This PR for the node sdk should be a good reference point: https://github.com/firebase/firebase-admin-node/pull/1044

@jeromegamez in terms of hardcoded urls Isn't the base_uri for auh the only thing we'd need to refactor? it doesn't look like it would be that hard to change that

**Comment by nelson6e65** [comment]
https://firebase.google.com/docs/emulator-suite/connect_auth#admin_sdks

I was able to use `FIREBASE_AUTH_EMULATOR_HOST` for auth.

I set `FIRESTORE_EMULATOR_HOST`, but then I get an error:

 ```
 Google\Cloud\Core\Exception\ServiceException

  {
    "message": "Established channel does not have a sufficient security level to transfer call credential.",
    "code": 16,
    "status": "UNAUTHENTICATED",
    "details": []
}
```
